### PR TITLE
[k8s] Add support for autoscaling kubernetes clusters

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Adds `autoscaler` field to config.yaml to allow the user to specify the autoscaler used in the underlying kubernetes cluster:
```
kubernetes:
  autoscaler: gke/karpenter/generic
```

Setting this field:
1. Disables gpu capacity checks during provisioning
2. Uses the appropriate label formatter for specifying GPU labels

When used in conjunction with `provision_timeout`, this allows users to bring in their scale-to-zero k8s clusters and use them with SkyPilot.

Currently supported autoscalers: GKE nodepools, Karpenter, any generic CA that can label new nodes with `skypilot.co/accelerator` (Related - https://github.com/skypilot-org/skypilot/issues/3432)

More context: https://docs.google.com/document/d/17LRYGCKDsY9AygAJbUIiQR4KkHZqSakL8G0SF7WNOR8/edit?usp=sharing

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Tested manually on a Karpenter cluster with autoscaling T4 GPU nodes
- [x] Tested manually on a GKE fixed cluster with 1x T4 GPU nodes 
- [x] Tested manually on a GKE autoscaling cluster with T4 GPU nodepool

### CODE REVIEW

File: docs/source/reference/config.rst [hunk 1]
```diff
@@ -313,6 +313,23 @@ Available fields and semantics:
...
+    # provision nodes with the required GPU capacity.
+    #
+    # Remember to set provision_timeout accordingly when using an autoscaler.
+    #
+    # Supported values: gke, karpenter, generic
+    #   gke: uses cloud.google.com/gke-accelerator label to identify GPUs on nodes
+    #   karpenter: uses karpenter.k8s.aws/instance-gpu-name label to identify GPUs on nodes
+    #   generic: uses skypilot.co/accelerator labels to identify GPUs on nodes
+    #
+    # Default: not set (no autoscaler, autodetect label format for GPU nodes)
```
Comment (Michaelvll) [inline docs/source/reference/config.rst:330]:
```suggestion
    # Default: null (no autoscaler, autodetect label format for GPU nodes)
```

File: docs/source/reference/config.rst [hunk 2]
```diff
@@ -313,6 +313,23 @@ Available fields and semantics:
...
+    # Kubernetes cluster. Setting this field disables pre-launch checks for
+    # GPU capacity in the cluster and SkyPilot relies on the autoscaler to
+    # provision nodes with the required GPU capacity.
+    #
+    # Remember to set provision_timeout accordingly when using an autoscaler.
+    #
+    # Supported values: gke, karpenter, generic
+    #   gke: uses cloud.google.com/gke-accelerator label to identify GPUs on nodes
+    #   karpenter: uses karpenter.k8s.aws/instance-gpu-name label to identify GPUs on nodes
+    #   generic: uses skypilot.co/accelerator labels to identify GPUs on nodes
```
Comment (Michaelvll) [inline docs/source/reference/config.rst:328]: Should we refer to the doc that sets up such labels?

File: sky/cli.py [hunk 3]
```diff
@@ -2998,23 +2998,30 @@ def _output():
 
         name, quantity = None, None
 
+        # Kubernetes specific bools
+        cloud_is_kubernetes = (cloud_obj is not None and
+                               cloud_obj.is_same_cloud(clouds.Kubernetes()))
```
Comment (Michaelvll) [inline sky/cli.py:3003]: nit:
```suggestion
        cloud_is_kubernetes = isinstance(cloud_obj, clouds.Kubernetes)
```

File: sky/cli.py [hunk 4]
```diff
@@ -3041,6 +3048,8 @@ def _output():
             else:
                 yield ('\n\nHint: use -a/--all to see all accelerators '
                        '(including non-common ones) and pricing.')
+                if cloud_is_kubernetes and kubernetes_autoscaling:
+                    yield kubernetes_utils.KUBERNETES_AUTOSCALER_NOTE
```
Comment (Michaelvll) [inline sky/cli.py:3052]: Does this mean when the `cloud` is not specified, i.e. None, we will not print the hint? Should we print the hint if `cloud is None` as well.

Also, in the show_all case, should we print this hint as well?

File: sky/provision/kubernetes/utils.py [hunk 5]
```diff
@@ -348,10 +372,30 @@ def get_gpu_label_key_value(acc_type: str, check_mode=False) -> Tuple[str, str]:
...
+            # early since we assume the cluster autoscaler will handle GPU
+            # node provisioning.
+            return '', ''
+        formatter = AUTOSCALER_TO_LABEL_FORMATTER.get(autoscaler_type)
+        if formatter is None:
+            # This should never be reached unless schema checks fail or
+            # AUTOSCALER_TO_LABEL_FORMATTER dict is not up to date.
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'Unsupported autoscaler type: {autoscaler_type}')
```
Comment (Michaelvll) [inline sky/provision/kubernetes/utils.py:396]: If it is not expected, an `assert` might be more accurate than raising an error?

File: sky/provision/kubernetes/utils.py [hunk 6]
```diff
@@ -1310,3 +1354,14 @@ def get_head_pod_name(cluster_name_on_cloud: str):
...
+
+
+def get_autoscaler_type(
+) -> Optional[kubernetes_enums.KubernetesAutoscalerType]:
+    """Returns the autoscaler type by reading from config"""
+    autoscaler_type = skypilot_config.get_nested(['kubernetes', 'autoscaler'],
+                                                 None)
+    if autoscaler_type is not None:
+        autoscaler_type = kubernetes_enums.KubernetesAutoscalerType.from_str(
+            autoscaler_type)
```
Comment (Michaelvll) [inline sky/provision/kubernetes/utils.py:1366]: nit: seems we have the schema check for the autoscaler_type already, should we just do `kubernetes_enums.KubernetesAutoscalerType(autoscaler_type)`?

**Review by Michaelvll — COMMENTED** [review]
Thanks for quickly adding this support @romilbhardwaj! Looks mostly good to me.

File: sky/cli.py (hunk 4 above)
Comment (romilbhardwaj) [reply sky/cli.py:3052]: ahh good point - fixed now! shown when cloud is not specified and also in show_all case.

**Review by Michaelvll — APPROVED** [review]
LGTM! Thanks @romilbhardwaj !

**Comment by romilbhardwaj** [comment]
Thanks @Michaelvll! Tested on karpenter and GKE nodepool autoscaling, merging now.

**Pull request merged**